### PR TITLE
add API for Gandi LiveDNS

### PR DESCRIPTION
**Pull request by fcrozat**
This adds support for new API (still in beta) available from Gandi

### CODE REVIEW

File: dnsapi/dns_gandi_livedns.sh [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#
+#Author: Frédéric Crozat <fcrozat@suse.com>
+#Report Bugs here: https://github.com/fcrozat/acme.sh
+#
+########  Public functions #####################
+
+GANDI_LIVEDNS_API="https://dns.beta.gandi.net/api/v5"
+
+#Usage: dns_gandi_livedns_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_gandi_livedns_add() {
```
Comment (Neilpang) [inline dnsapi/dns_gandi_livedns.sh]: why not name it `dns_gandi` instead of `dns_gandi_livedns`  ?

Do you mean gandi has other dns system ?

Comment (fcrozat) [inline dnsapi/dns_gandi_livedns.sh]: Gandi DNS is redoing their DNS infrastructure. Their old infrastructure (and API) was making DNS-01 validation very hard to implement (minimum of 20mins for DNS zone push, and sometime, it was even longer). Their new infrastructure (called LiveDNS) is only available as beta, uses a REST api and supports "instant" push of DNS changes. That is why I named it gandi_livedns.. I think they will phase out their "old" API in the near future but I don't know when.

But I can rename it to gandi instead, of course.

Comment (Neilpang) [inline dnsapi/dns_gandi_livedns.sh]: ok, keep it there.

File: dnsapi/dns_gandi_livedns.sh [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  fulldomain=$1
+  txtvalue=$2
+
+  if [ -z "$GANDI_LIVEDNS_KEY" ]; then
+      _err "No API key specifed for Gandi LiveDNS."
+      _err "Create your key and export it as GANDI_LIVEDNS_KEY"
+      return 1
+  fi
+
```
Comment (Neilpang) [inline dnsapi/dns_gandi_livedns.sh]: please save the gandi api key, the future auto renewal will use the key again.

```sh
_saveaccountconf GANDI_LIVEDNS_KEY   "$GANDI_LIVEDNS_KEY"
```

File: dnsapi/dns_gandi_livedns.sh [hunk 3]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env sh
+
+# Gandi LiveDNS v5 API
+# http://doc.livedns.gandi.net/
+# currently under beta
+#
+# Requires GANDI API KEY set in GANDI_LIVEDNS_KEY set as environment variable
+#
+#Author: Frédéric Crozat <fcrozat@suse.com>
+#Report Bugs here: https://github.com/fcrozat/acme.sh
```
Comment (Neilpang) [inline dnsapi/dns_gandi_livedns.sh]: add your repository here to report bugs.

Comment (fcrozat) [inline dnsapi/dns_gandi_livedns.sh]: hmm, it is my repository (fcrozat/acme.sh)

Comment (Neilpang) [inline dnsapi/dns_gandi_livedns.sh]: Sorry, my eyes broken. :)

File: dnsapi/dns_gandi_livedns.sh [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  if ! _get_root "$fulldomain"; then
+    _err "invalid domain"
+    return 1
+  fi
+  _debug fulldomain "$fulldomain"
+  _debug txtvalue "$txtvalue"
+  _debug domain "$_domain"
+  _debug sub_domain "$_sub_domain"
+
+  _gandi_livedns_rest POST "domains/$_domain/records/$_sub_domain/TXT" "{\"rrset_ttl\": 300, \"rrset_values\":[\"$txtvalue\"]}"
```
Comment (baloo) [inline dnsapi/dns_gandi_livedns.sh]: You may use PUT here which will overwrite any existing TXT record here. This is also idempotent.

Comment (baloo) [inline dnsapi/dns_gandi_livedns.sh]: The ttl is also optional, it will default to the soa value

**Comment by Neilpang** [comment]
Please fix the CI format error https://travis-ci.org/Neilpang/acme.sh/builds/205790181

File: dnsapi/dns_gandi_livedns.sh (hunk 4 above)
Comment (fcrozat) [inline dnsapi/dns_gandi_livedns.sh]: POST vs PUT: the specific API in the doc for creating a record was mentionning POST by default. I'll switch to PUT.
For TTL, draft-ietf-acme-acme-01 is suggesting a TTL of 300.

**Comment by Neilpang** [comment]
I was not able to use this api.
Do I need to do something to enable this live dns api ?

**Comment by fcrozat** [comment]
You need to migrate your account to Gandi v5 and switch to their new DNS servers: https://news.beta.gandi.net/en/10215-Gandi-s-new-platform-it-s-here and

**Comment by Neilpang** [comment]
trying.

**Comment by Neilpang** [comment]
Merged.  Thank you.

**Pull request merged**